chrome/content/zotero/xpcom/preferencePanes.js

```diff
@@ -0,0 +1,131 @@
+	
+	Zotero is free software: you can redistribute it and/or modify
+	it under the terms of the GNU Affero General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+	
+	Zotero is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU Affero General Public License for more details.
+
+	You should have received a copy of the GNU Affero General Public License
+	along with Zotero.  If not, see <http://www.gnu.org/licenses/>.
+	
+	***** END LICENSE BLOCK *****
+*/
+
+/**
+ * Manages preference panes.
+ */
+Zotero.PreferencePanes = {
+	builtInPanes: Object.freeze([
+		{
+			id: 'zotero-prefpane-general',
+			label: 'zotero.preferences.prefpane.general',
+			image: 'chrome://zotero/skin/prefs-general.png',
+			src: 'chrome://zotero/content/preferences/preferences_general.xhtml',
+			scripts: ['chrome://zotero/content/preferences/preferences_general.js'],
+			defaultXUL: true,
+		},
+		{
+			id: 'zotero-prefpane-sync',
+			label: 'zotero.preferences.prefpane.sync',
+			image: 'chrome://zotero/skin/prefs-sync.png',
+			src: 'chrome://zotero/content/preferences/preferences_sync.xhtml',
+			scripts: ['chrome://zotero/content/preferences/preferences_sync.js'],
+			defaultXUL: true,
+		},
+		{
+			id: 'zotero-prefpane-export',
+			label: 'zotero.preferences.prefpane.export',
+			image: 'chrome://zotero/skin/prefs-export.png',
+			src: 'chrome://zotero/content/preferences/preferences_export.xhtml',
+			scripts: ['chrome://zotero/content/preferences/preferences_export.js'],
+			defaultXUL: true,
+		},
+		{
+			id: 'zotero-prefpane-cite',
+			label: 'zotero.preferences.prefpane.cite',
+			image: 'chrome://zotero/skin/prefs-styles.png',
+			src: 'chrome://zotero/content/preferences/preferences_cite.xhtml',
+			scripts: ['chrome://zotero/content/preferences/preferences_cite.js'],
+			defaultXUL: true,
+		},
+		{
+			id: 'zotero-prefpane-advanced',
+			label: 'zotero.preferences.prefpane.advanced',
+			image: 'chrome://zotero/skin/prefs-advanced.png',
+			src: 'chrome://zotero/content/preferences/preferences_advanced.xhtml',
+			scripts: ['chrome://zotero/content/preferences/preferences_advanced.js'],
+			defaultXUL: true,
+		},
+		{
+			id: 'zotero-subpane-reset-sync',
+			parent: 'zotero-prefpane-sync',
+			label: 'zotero.preferences.subpane.resetSync',
+			src: 'chrome://zotero/content/preferences/preferences_sync_reset.xhtml',
+			scripts: ['chrome://zotero/content/preferences/preferences_sync.js'],
+			defaultXUL: true,
+		}
+	]),
+
+	pluginPanes: [],
+
+	/**
+	 * Register a pane to be displayed in the preferences. The pane XHTML (`src`)
+	 * is loaded as a fragment, not a full document, with XUL as the default
+	 * namespace and (X)HTML tags available under `html:`.
+	 *
+	 * Plugins must call #unregister(id) in their shutdown() method.
```

Comment:
Since we're requiring a `pluginID`, any reason we shouldn't just unregister the pane ourselves (and say here that the pane will be automatically unregistered if the plugin is disabled)? We can unregister from the `onUninstalled()` handler in `Zotero.Plugins`.

Reply:
So that we avoid a circular dependency between `Zotero.Plugins` and `Zotero.PreferencePanes`, how about adding support for lifecycle observers to `Zotero.Plugins`? We could either allow observers to implement any of the methods on `_addonObserver`, or, if we'd like to avoid exposing the Mozilla-specific API, just `onInstall`, `onStartup`, `onShutdown`, and `onUninstall` (like the plugins themselves get).

Reply:
I dunno — what's the specific concern?

If anything I think we'd want to expose the same events the plugins get, but I don't think it's necessary for this vs. a one-line call to `Zotero.PreferencePanes.unregister()` in `onUninstalled()`. I think the main reason to do it would be if we think plugins want the ability to monitor for events from other plugins, which is not impossible — there might be some inter-plugin customizations already. They could just use `AddonManager.addAddonListener()` directly but providing our own listeners would be better. I think we could ignore this for now, but if you want to do that, that's fine.

Reply:
`unregister()` wants a pane ID, not a plugin ID, so we'd need to expose a method to unregister by plugin ID, and that doesn't seem useful beyond being called from `Zotero.Plugins`. The interoperability case is also worth supporting. So why not, I guess. Let me know if what I pushed looks OK.